# Dlaczego wasz model jest lepszy od innych? 


Otrzymaliśmy 92% trafności na danych testowych. Uważamy, że model jest wystarczająco dobry, żeby móc go zaimplementować

# Jak to sprawdziliście? 


Model przetrenowaliśmy na oczyszczonych danych (9826 pomiarów). Zawsze otrzymaliśmy skuteczność około 91%

# Dlaczego podjęliście takie a nie inne decyzje?


Chcieliśmy uprościć problem jak najbardziej. Dlatego podczas oczyszczania danych usunęliśmy wszystkie wiersze, które zawierały NaNy - błędne pomiary nie powinny być interpretowane przez model. Następnie wybraliśmy tylko wartości określone w kolumnach jako średnia (mean) lub odchylenie standardowe (std) - czyli najważniejsze dla nas wartości. Dzięki temu zmnieszyliśmy wymiarowość danych do 66 cech. Żeby sieć nie uczyła się też na danych, które zostały błędnie odczytane, obliczyliśmy dla każdej osoby (subject) w ramach każdej aktywności, wartości średnie dla wspomnianych wcześniej cech. Dało to nam 160 pomiarów treningowych. To wystarczyło sieci neuronowej na bardzo efektywną naukę.

Użyliśmy funkcji aktywacji softmax na końcu ze względu na jej charakter. Zwraca nam ona listę prawdopodobieństw, która sumuje się do 1. Można dzięki temu interpretować, które aktywności są najbardziej podobne do danego pomiaru, a które mniej

Kryteria oceny:
1. Czyszczenie danych i redukcja wymiarowości. (10 pkt)*
2. Dokładność modelu. (25 pkt)
3. Zrozumienie i interpretacja wyników. (35 pkt)*
4. Prezentacja i użyteczność biznesowa. (20 pkt)

## Wstępna analiza

Wczytujemy plik csv używając biblioteki pandas jako DataFrame, upewniamy się że widoczne będą wszystkie kolumny z danymi, analizujemy wielkość i naturę zbioru danych.

In [1]:
import pandas as pd
df = pd.read_csv('data/final_train.csv')
pd.set_option('display.max_columns', None)

In [2]:
df.shape

(11086, 573)

In [3]:
df.sample(6)

Unnamed: 0  Activity  angle(X,gravityMean)  angle(Y,gravityMean)  \
1849        1849  STANDING             -1.568817             -1.017461   
655          655  STANDING             -1.573922             -0.797537   
5211        5211   WALKING             -1.572888             -0.850793   
9984        9984    LAYING             -1.548333             -0.139628   
7825        7825   SITTING             -1.571599             -0.775193   
9034        9034    LAYING             -1.549744             -0.079625   

      angle(Z,gravityMean)  angle(tBodyAccJerkMean),gravityMean)  \
1849             -2.444467                             -1.829731   
655              -2.449280                             -1.844928   
5211             -2.449691                             -1.840999   
9984             -2.380889                             -1.915239   
7825             -2.436326                             -1.838110   
9034             -2.388229                             -1.865817   

      angle(tBodyAccMean,gravity)  angle(tBodyGyroJerkMean,gravityMean)  \
1849                     0.115072                             -1.614248   
655                      0.139001                             -0.048197   
5211                    -0.276919                              0.265815   
9984                     0.067613                             -0.887342   
7825                     0.057135                              0.243189   
9034                     0.069862                             -1.544690   

      angle(tBodyGyroMean,gravityMean)  energy-mean()  \
1849                          3.008947      -8.945338   
655                           5.755578      -5.655270   
5211                          8.075734     -11.953896   
9984                          6.007361     -11.801947   
7825                          5.820848      -7.198437   
9034                          3.335026      -6.303845   

      fBodyAcc-bandsEnergy()-1,16  fBodyAcc-bandsEnergy()-1,16.1  \
1849                     3.623850                       6.158873   
655                      3.624113                       6.158605   
5211                     2.757560                       5.542987   
9984                     3.621984                       6.158630   
7825                     3.624008                       6.158670   
9034                     3.617229                       6.129565   

      fBodyAcc-bandsEnergy()-1,16.2  fBodyAcc-bandsEnergy()-1,24  \
1849                       2.495142                     5.777323   
655                        2.495181                     5.777383   
5211                       2.551391                     5.545049   
9984                       2.495276                     5.776843   
7825                       2.495638                     5.777370   
9034                       2.499364                     5.775270   

      fBodyAcc-bandsEnergy()-1,24.1  fBodyAcc-bandsEnergy()-1,24.2  \
1849                       3.315233                      -1.352859   
655                        3.314339                      -1.352859   
5211                       1.305766                      -1.353453   
9984                       3.314383                      -1.352860   
7825                       3.314580                      -1.352863   
9034                       3.208710                      -1.352917   

      fBodyAcc-bandsEnergy()-1,8  fBodyAcc-bandsEnergy()-1,8.1  \
1849                   -1.537924                      2.845908   
655                    -1.537951                      2.845765   
5211                   -1.376737                      2.490582   
9984                   -1.537483                      2.845718   
7825                   -1.537932                      2.845694   
9034                   -1.536867                      2.828492   

      fBodyAcc-bandsEnergy()-1,8.2  fBodyAcc-bandsEnergy()-17,24  \
1849                      8.619072                     -2.232494   
655                       8.619025      

W trakcie analizy dochodzimy do kilku wniosków, mianowicie:

In [4]:
pd.unique(df['Activity'])

array(['STANDING', 'WALKING_UPSTAIRS', 'LAYING', 'WALKING', 'SITTING',
       'WALKING_DOWNSTAIRS'], dtype=object)

1. Mamy 6 aktywności, które nasz model ma rozpoznawać,

In [5]:
df.subject.nunique()

30

2. W badaniu brało udział 30 uczestników,

3. Do trenowania modelu klasyfikacji definitywnie nie potrzebujemy wszystkich cech. Skupiamy się na tych, które zawierają średnią i odchylenie standardowe.

## Oczyszczanie danych

Zanim zaczniemy czyścić zbiór, wyciągamy z niego dane o uczestnikach i aktywnościach, to przyda nam się potem.

In [6]:
subject = df['subject']
activity = df['Activity']

Na początku wybieramy wszystkie kolumny zawierające średnią lub odchylenie standardowe wszystkich z cech.

In [7]:
mean_cols = [col for col in df.columns if 'mean' in col or 'Mean' in col]
std_cols = [col for col in df.columns if 'std' in col or 'Std' in col]

df = pd.concat([df[std_cols], df[mean_cols]], axis = 1)
df.shape

(11086, 87)

In [8]:
573-87, (573-87)/573

(486, 0.8481675392670157)

W ten sposób udało nam się pozbyć 486 cech, co stanowi prawie 85% całego zbioru.

Aktualnie nasz zbiór wygląda tak:

In [9]:
df.sample(4)

fBodyAcc-std()-X  fBodyAcc-std()-Y  fBodyAcc-std()-Z  \
5831           0.797941         -1.267550          4.182218   
5345          -1.223073         -3.158632          3.912448   
2428           0.799663         -1.348191          4.212089   
10448         -0.074151         -3.102851          3.986716   

       fBodyAccJerk-std()-X  fBodyAccJerk-std()-Y  fBodyAccJerk-std()-Z  \
5831              -4.028178              0.261975              2.359286   
5345              -5.525108              2.343067              1.473819   
2428              -4.034683              0.277929              2.364314   
10448             -5.402266              2.237942              1.923581   

       fBodyAccMag-std()  fBodyBodyAccJerkMag-std()  \
5831           -0.816908                  -3.016575   
5345           -2.797628                  -2.847767   
2428           -0.832648                  -3.015946   
10448          -1.623248                  -2.873671   

       fBodyBodyGyroJerkMag-std()  fBodyBodyGyroMag-std()  fBodyGyro-std()-X  \
5831                    -6.016181                4.832826           3.221559   
5345                    -6.075579                4.619765           1.194675   
2428                    -6.016224                4.839664           3.250277   
10448                   -6.060949                4.744567           1.956542   

       fBodyGyro-std()-Y  fBodyGyro-std()-Z  tBodyAcc-std()-X  \
5831            0.804087           4.685762          6.346211   
5345            0.740399           4.213907          5.690689   
2428            0.812255           4.709953          6.346384   
10448           0.761174           4.374398          6.030422   

       tBodyAcc-std()-Y  tBodyAcc-std()-Z  tBodyAccJerk-std()-X  \
5831           3.884976          3.394077              6.312624   
5345           3.738668          2.671849              5.187333   
2428           3.879430          3.455001              6.307238   
10448          3.745401          2.928167              5.270058   

       tBodyAccJerk-std()-Y  tBodyAccJerk-std()-Z  tBodyAccJerkMag-std()  \
5831              -0.522479              3.924207               2.310968   
5345              -2.401673              3.330657               3.083971   
2428              -0.539922              3.926187               2.314490   
10448             -2.361726              3.630036               2.995065   

       tBodyAccMag-std()  tBodyGyro-std()-X  tBodyGyro-std()-Y  \
5831            1.625686           5.283754           5.552676   
5345           -1.287540           3.295311           3.772891   
2428            1.612246           5.309700           5.719632   
10448           0.311126           4.019369           4.397375   

       tBodyGyro-std()-Z  tBodyGyroJerk-std()-X  tBodyGyroJerk-std()-Y  \
5831            1.774669              -4.156204              -0.589930   
5345            1.242333              -3.156759              -0.494562   
2428            1.795215              -4.162492              -0.590106   
10448           1.450602              -3.413486              -0.538470   

       tBodyGyroJerk-std()-Z  tBodyGyroJerkMag-std()  tBodyGyroMag-std()  \
5831                0.852433               -2.665807            5.710371   
5345                0.854299               -2.337315            3.988794   
2428                0.852426               -2.666165            5.760531   
10448               0.853669               -2.448556            4.919735   

       tGravityAcc-std()-X  tGravityAcc-std()-Y  tGravityAcc-std()-Z  \
5831              3.742629             1.438680             4.079544   
5345              3.693668             1.272762             4.128248   
2428              3.721022             1.347543             4.101103   
10448             3.726561             1.391572             4.143428   

       tGravityAccMag-std()  angle(X,gravityMean)  angle(Y,gravityMean)  \
5831               2.026792             -1.574933             -0.606069   
5

Dodatkowo zdecydowaliśmy się pozbyć kolumn, które zawierają słowa kluczowe: 'angle', 'energy' i 'meanFreq'

In [10]:
angle_cols = [col for col in df.columns if 'angle' in col or 'Angle' in col]
energy_cols = [col for col in df.columns if 'energy' in col or 'Energy' in col]
meanFreq_cols = [col for col in df.columns if 'meanFreq' in col]

drop_cols = angle_cols + energy_cols + meanFreq_cols
df = df.drop(drop_cols, axis=1)

In [11]:
df.shape

(11086, 66)

Ostatecznie nasz zbiór ma 66 cech.

Dodajemy kolumny uczestnika i aktywności.

In [12]:
df.insert(0, 'Activity', activity)
df.insert(0, 'Subject', subject)

Pozbywamy się niekompletnych wierszy.

In [ ]:
df = df.dropna()

Ostatecznie zdecydowaliśmy się zgrupować wszystkich uczestników badania razem z aktywnościami, w których brali udział

In [ ]:
grouped = df.groupby(['Subject', 'Activity'])

Wyciągamy średnią z wszystkich cech, dla każdego uczestnika względem każdej z badanych aktywności.

In [ ]:
df = grouped.mean().reset_index()

I tak prezentują się nasze oczyszczone dane:

In [ ]:
df.head(10)

In [ ]:
d